src/signature_mahalanobis_knn/sig_mahal_knn.py

```diff
@@ -5,117 +5,179 @@
+                raise ValueError(msg)
+
+            from sktime.transformations.panel.signature_based import (
+                SignatureTransformer,
+            )
+
+            self.signature_transform = SignatureTransformer(
+                **kwargs,
+            )
+
+            # compute signatures
+            sigs = Parallel(n_jobs=self.n_jobs)(
+                delayed(self.signature_transform.fit_transform)(X[i])
+                for i in range(len(X))
+            )
+            self.signatures = pd.concat(sigs)
+        else:
+            self.signatures = signatures
+
+        # fit mahalanobis distance for the signatures
         mahal_distance = Mahalanobis()
-        mahal_distance.fit(sigs)
+        mahal_distance.fit(self.signatures)
 
+        # fit knn for the mahalanobis distance
         knn = NearestNeighbors(
-            metric=mahal_distance.distance, n_jobs=self.n_jobs, algorithm="auto"
+            metric=mahal_distance.distance,
+            n_jobs=self.n_jobs,
+            algorithm=knn_algorithm,
         )
-        knn.fit(sigs)
+        knn.fit(self.signatures)
         self.knn = knn
 
-    def conformance(self, X):
+    def conformance(
+        self,
+        X: np.ndarray | None = None,
+        signatures: np.ndarray | None = None,
+    ) -> np.ndarray:
         """
-        __description__
-
-        :param X: Must support index operation X[i] where each X[i] returns a data point
-
-        :return:
+        Compute the conformance scores for the data points either passed in
+        directly as X or the signatures of the data points in signatures.
+        If signatures is not provided, then X must be provided
+        to compute the signatures.
+        If signatures is provided, then X is ignored.
+
+        Must call fit() method first.
+
+        Parameters
+        ----------
+        X : np.ndarray | None, optional
+            Data points, by default None.
+            Must support index operation X[i] where
+            each X[i] returns a data point in the corpus.
+        signatures : np.ndarray | None, optional
+            Signatures of the data points, by default None.
+            Must support index operation X[i] where
+            each X[i] returns a data point in the corpus.
+
+        Returns
+        -------
+        np.ndarray
+            Conformance scores for data points provided.
         """
-        sigs = Parallel(n_jobs=self.n_jobs)(
-            delayed(self.signature_transform.fit_transform)(X[i]) for i in range(len(X))
+        if self.knn is None:
+            msg = "Must fit the model first"
+            raise ValueError(msg)
+
+        if signatures is None:
+            if X is None:
+                msg = "Either X or signatures must be provided"
+                raise ValueError(msg)
+            if self.signature_transform is None:
+                msg = "Must fit the model first"
+                raise ValueError(msg)
```

Comment:
There are code duplications between the preceding 10 lines and line 63-73 (approximately), is it possible to write it as a separate logic for better maintainability and readability?

Reply:
sure, will do

Reply:
That would be great thanks.